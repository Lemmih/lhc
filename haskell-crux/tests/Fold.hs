module Fold where

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr fn acc [] = acc
foldr fn acc (x:xs) = fn x (foldr fn acc xs)

foldl :: (b -> a -> b) -> b -> [a] -> b
foldl fn acc [] = acc
foldl fn acc (x:xs) = foldl fn (fn acc x) xs
