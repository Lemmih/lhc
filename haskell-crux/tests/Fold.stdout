Fold.foldr : ∀ a b. (a →  b →  b) →  b →  [a] →  b
Fold.foldr =
  λ @.a^1:* @.b^2:* Fold.fn^3:(a →  b →  b) Fold.acc^4:b arg^5:[a] → 
  case arg^5 of
    scrut^6:[a]
    LHC.Prim.Nil  → 
      Fold.acc^4
    LHC.Prim.Cons Fold.x^7:a Fold.xs^8:[a] → 
      Fold.fn^3 Fold.x^7 (Fold.foldr
                  @.a^1 @.b^2 Fold.fn^3 Fold.acc^4
          Fold.xs^8)
Fold.foldl : ∀ a b. (b →  a →  b) →  b →  [a] →  b
Fold.foldl =
  λ @.a^9:* @.b^10:* Fold.fn^11:(b →  a →  b) Fold.acc^12:b arg^13:[a] → 
  case arg^13 of
    scrut^14:[a]
    LHC.Prim.Nil  → 
      Fold.acc^12
    LHC.Prim.Cons Fold.x^15:a Fold.xs^16:[a] → 
      Fold.foldl @.a^9 @.b^10
            Fold.fn^11 (Fold.fn^11
              Fold.acc^12 Fold.x^15)
        Fold.xs^16
